store/index.go

```diff
@@ -0,0 +1,324 @@
+
+	indexValue := make(keyList, 0)
+
+	if err != ds.ErrNotFound {
+		if index.Unique && !delete {
+			return ErrUniqueExists
+		}
+	}
+
+	if data != nil {
+		err = DefaultDecode(data, &indexValue)
+		if err != nil {
+			return err
+		}
+	}
+
+	if delete {
+		indexValue.remove(key)
+	} else {
+		indexValue.add(key)
+	}
+
+	if len(indexValue) == 0 {
+		return tx.Delete(indexKey)
+	}
+
+	iVal, err := DefaultEncode(indexValue)
+	if err != nil {
+		return err
+	}
+
+	return tx.Put(indexKey, iVal)
+}
+
+// keyList is a slice of unique, sorted keys([]byte) such as what an index points to
+type keyList [][]byte
+
+func (v *keyList) add(key ds.Key) {
+	b := key.Bytes()
+	i := sort.Search(len(*v), func(i int) bool {
+		return bytes.Compare((*v)[i], b) >= 0
+	})
+
+	if i < len(*v) && bytes.Equal((*v)[i], b) {
+		// already added
+		return
+	}
+
+	*v = append(*v, nil)
+	copy((*v)[i+1:], (*v)[i:])
+	(*v)[i] = b
+}
+
+func (v *keyList) remove(key ds.Key) {
+	b := key.Bytes()
+	i := sort.Search(len(*v), func(i int) bool {
+		return bytes.Compare((*v)[i], b) >= 0
+	})
+
+	if i < len(*v) {
+		copy((*v)[i:], (*v)[i+1:])
+		(*v)[len(*v)-1] = nil
+		*v = (*v)[:len(*v)-1]
+	}
+}
+
+func (v *keyList) in(key ds.Key) bool {
+	b := key.Bytes()
+	i := sort.Search(len(*v), func(i int) bool {
+		return bytes.Compare((*v)[i], b) >= 0
+	})
+
+	return (i < len(*v) && bytes.Equal((*v)[i], b))
+}
+
+// FilterQuery provides a Query Filter function for querying JSON data
+type FilterQuery struct {
+	query *JSONQuery
+}
+
+// Filter is the primary Filter function
+func (f FilterQuery) Filter(e query.Entry) bool {
+	val := make(map[string]interface{})
+	if err := json.Unmarshal(e.Value, &val); err != nil {
+		return false
+	}
+	ok, err := f.query.matchJSON(val)
+	return ok && err == nil
+}
+
+type iterator struct {
+	nextKeys func() ([]ds.Key, error)
+	txn      ds.Txn
+	query    *JSONQuery
+	err      error
+	keyCache []ds.Key
+	iter     query.Results
+}
+
+func newIterator(txn ds.Txn, baseKey ds.Key, q *JSONQuery) *iterator {
+	i := &iterator{
+		txn:   txn,
+		query: q,
+	}
+	// Key field or index not specified, pass thru to base 'iterator'
+	if q.Index == "" {
+		dsq := query.Query{
+			Prefix: baseKey.String(),
+			Filters: []query.Filter{
+				FilterQuery{
+					query: q,
+				},
+			},
+		}
+		i.iter, i.err = txn.Query(dsq)
+		i.nextKeys = func() ([]ds.Key, error) {
+			return nil, nil
+		}
+		return i
+	}
+
+	// indexed field, get keys from index
+	indexKey := indexPrefix.Child(baseKey).ChildString(q.Index)
+	dsq := query.Query{
+		Prefix: indexKey.String(),
+	}
+	i.iter, i.err = txn.Query(dsq)
+	first := true
+	i.nextKeys = func() ([]ds.Key, error) {
+		var nKeys []ds.Key
+
+		for len(nKeys) < iteratorKeyMinCacheSize {
+			result, ok := i.iter.NextSync()
+			if !ok {
+				if first {
+					return nil, ErrNoIndexFound
+				}
+				return nKeys, result.Error
+			}
+			first = false
+			// result.Key contains the indexed value, extract here first
+			key := ds.RawKey(result.Key)
```

Comment:
If we do have an index, we extract the value, perform the comparison, and then start returning the keys underlying that index value.

Reply:
We can do this in batches (iteratorKeyMinCacheSize ) if we want, though in practice, this probably isn't needed some of the time given we'll return the results as an array of strings in the end.